proposals/NNNN-dxil-vectors.md

```diff
@@ -0,0 +1,183 @@
+      i32,                  ; opcode
+      %dx.types.Handle,     ; resource handle
+      i32,                  ; coordinate c0 (index)
+      i32,                  ; coordinate c1 (elementOffset)
+      i8,                   ; mask
+      i32,                  ; alignment
+  )
+```
+
+The return struct contains a single vector and a single integer representing mapped tile status.
+
+```asm
+  %dx.types.ResRet.v[NUM][TY] = type { vector<TYPE, NUM>, i32 }
+```
+
+Here and hereafter, `NUM` is the number of elements in the loaded vector, `TYPE` is the element type name,
+ and `TY` is the corresponding abbreviated type name (e.g. `i64`, `f32`).
+
+#### Elementwise intrinsics
+
+A selection of elementwise intrinsics are given additional native vector forms.
+Elementwise intrinsics are those that perform their calculations irrespective of the location of the element
+ in the vector or matrix arguments except insofar as that position corresponds to those of the other elements
+ that might be used in the individual element calculations.
+An elementwise intrinsic `foo` that takes scalar or vector arguments could theoretically implement its vector version using a simple loop and the scalar intrinsic variant.
+
+```c++
+vec<TYPE, NUM> foo(vec<TYPE, NUM> a, vec<TYPE, NUM> b) {
+  vec<TYPE, NUM> ret;
+  for (int i = 0; i < NUM; i++)
+    ret[i] = foo(a[i], b[i]);
+}
+```
+  
+For example, `fma` is an elementwise intrinsic because it multiplies or adds each element of its argument vectors,
+ but `cross` is not because it performs an operation on the vectors as units,
+ pulling elements from different locations as the operation requires.
+
+The elementwise intrinsics that have native vector variants represent the
+ unary, binary, and tertiary generic operations:
+
+```asm
+ <[NUM] x [TYPE]> @dx.op.unary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1)
+ <[NUM] x [TYPE]> @dx.op.binary.v[NUM][[TY]](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2)
+ <[NUM] x [TYPE]> @dx.op.tertiary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2, <[NUM] x [TYPE]> operand3)
+```
+
+The only opcodes allowed with vector variants are:
+
+* Unary
+  * Exp
+  * Htan
+  * Atan
+  * Log
+* Binary
+  * FMin
+  * FMax
+* Tertiary
+  * Fma
+
+Unsupported DXIL intrinsics will continue to operate on scalarized representations even if those scalars
+ are extracted from native vectors.
+
+### Potential Changes to DXIL Consumers
+
+As this removes no existing DXIL features, the former representation of vectors is still valid.
+However, DXIL consumers may expect native vectors where they are supported and may misinterpret
+ vectors scalarized into arrays as being native arrays.
+This is unlikely to produce any faulty results, but may miss some optimizations.
+
+As DXIL with native vectors might be linked to create a DXIL shader without that support,
+ some additional scalarization might be necessary when linking in such cases.
```

Comment:
Do we support linking code built for different shader models?  I'd have thought we'd require the SM's to match when linking!

Reply:
Yes, I agree this is an implementation detail that doesn't belong in this spec.  We might want an infra spec to track details like this.